sematic/db/models/resolution.py

```diff
@@ -0,0 +1,177 @@
+    in one of the runs involved in it.
+
+    States
+    ------
+    CREATED:
+        The Resolution is in the DB, but k8s has not been asked to execute it yet.
+    SCHEDULED:
+        K8s (or the local process, for non-detached) has been asked to execute the
+        resolution, but the code has not started executing for it yet.
+    RUNNING:
+        The code for the resolution is executing.
+    FAILED:
+        There was an error in the resolution itself, NOT necessarily in the runs that it's
+        managing. The resolution may have started getting too many 500s from the Sematic
+        server, for example.
+    COMPLETE:
+        The resolution is done, and it will do no more work. This status may be used even
+        if the root run failed, so long as it failed due to some issue in the Sematic
+        func execution and not for some other reason. It may also be used if the root run
+        was canceled, so long as the cancellation was exited cleanly.
+    """
+
+    CREATED = "CREATED"
+    SCHEDULED = "SCHEDULED"
+    RUNNING = "RUNNING"
+    FAILED = "FAILED"
+    COMPLETE = "COMPLETE"
+
+    @classmethod
+    def is_allowed_transition(
+        cls,
+        from_status: Union["ResolutionStatus", str],
+        to_status: Union["ResolutionStatus", str],
+    ) -> bool:
+        if not isinstance(from_status, ResolutionStatus):
+            from_status = ResolutionStatus[from_status]
+        if not isinstance(to_status, ResolutionStatus):
+            to_status = ResolutionStatus[to_status]
+        return to_status in _ALLOWED_TRANSITIONS[from_status]
+
+
+_ALLOWED_TRANSITIONS = {
+    ResolutionStatus.CREATED: {ResolutionStatus.SCHEDULED, ResolutionStatus.FAILED},
+    ResolutionStatus.SCHEDULED: {ResolutionStatus.RUNNING, ResolutionStatus.FAILED},
+    ResolutionStatus.RUNNING: {ResolutionStatus.COMPLETE, ResolutionStatus.FAILED},
+    ResolutionStatus.COMPLETE: {},
+    ResolutionStatus.FAILED: {},
+}
+
+
+@unique
+class ResolutionKind(Enum):
+    """The kind of compute used to execute the resolution"""
+
+    LOCAL = "LOCAL"  # for non-detached mode
+    KUBERNETES = "KUBERNETES"  # for detached mode
+
+
+class Resolution(Base, JSONEncodableMixin):
+    """Represents a session of a resolver
+
+    Attributes
+    ----------
+    root_id:
+        The id of the root run which this resolution is resolving.
+    status:
+        The state of the resolver session, see ResolutionStatus.
+    kind:
+        The kind of resolver session (ex: on k8s or not).
+    docker_image_uri:
+        The docker image URI for the resolution. May be null when
+        doing a non-detached (local) resolution
+    settings_env_vars:
+        The Sematic settings from the user's environment for the user
+        who launched this resolution.
+    """
+
+    __tablename__ = "resolutions"
+
+    root_id: str = Column(
+        types.String(),
+        nullable=False,
+        primary_key=True,
+        default=lambda: uuid.uuid4().hex,
```

Comment:
Isn't this set by the client?

Reply:
yeah, I copied this from note and that's now it's id was set. Happy to remove it though.